tests/publiccloud/ipa.pm

```diff
@@ -77,61 +74,90 @@ sub run {
+        assert_script_run("wget " . get_required_var("PUBLIC_CLOUD_IMAGE_URL") . " -O " . $image_name, timeout => 300);
+
+        #Write ec2utils configuration, needed by ec2uploadimg
+        assert_script_run("echo -e '[region-eu-central-1]\\nami = ami-bc5b48d0\\ninstance_type = t2.micro\\naki_i386 = aki-3e4c7a23\\n"
+              . "aki_x86_64 = aki-184c7a05\\ng2_aki_x86_64 = aki-e23f09ff\\nuser = ec2-user\\n' >> ~/.ec2utils.conf");
+
+        assert_script_run(
+            "ec2uploadimg --access-id '"
+              . get_required_var('PUBLIC_CLOUD_KEY_ID')
+              . "' -s '"
+              . get_required_var('PUBLIC_CLOUD_KEY_SECRET') . "' "
+              . "--backing-store ssd "
+              . "--grub2 "
+              . "--machine 'x86_64' "
+              . "-n '$image_name' "
+              . (($image_name =~ /hvm/i) ? "--virt-type hvm --sriov-support " : "--virt-type para ")
+              . "--verbose "
+              . "--regions 'eu-central-1' "
+              . "--ssh-key-pair 'QA_SSH_KEY' "
+              . "--private-key-file 'QA_SSH_KEY.pem' "
+              . "-d 'OpenQA tests' "
+              . "'$image_name'",
+            timeout => 1200
+        );
         $ami_id = find_ami($image_name);
+        die "Failed on uploading $image_name" unless $ami_id;
     }
 
-
     #Prestart instance, cause IPA might use the wrong security group
     my ($instance_id)
       = script_output("aws ec2 run-instances --image-id $ami_id --instance-type t2.large --key-name QA_SSH_KEY --security-group-ids qa_secgroup")
       =~ /"InstanceId":\s*"([^"]+)"/;
 
-    # TODO download latest IPA tests to ~/ipa/tests/
+    # download latest IPA tests
+    #    assert_script_run("git clone -q --depth 1 https://github.com/SUSE/ipa.git ipa_repo");
+    #    assert_script_run("ln -s ipa_repo/usr/share/lib/ipa ipa");
 
-    assert_script_run("ipa test ec2 "
+    # Create some folders, ipa will need them
+    assert_script_run("mkdir -p ~/ipa/tests/");
+    assert_script_run("mkdir -p .config/ipa");
+    assert_script_run("touch .config/ipa/config");
+    assert_script_run("ipa list");
+
+    assert_script_run(
+            "ipa test ec2 "
           . "--access-key-id '"
           . get_required_var('PUBLIC_CLOUD_KEY_ID') . "' "
           . "--secret-access-key '"
           . get_required_var('PUBLIC_CLOUD_KEY_SECRET') . "' "
           . "-D 'IPA test $image_name' "
           . "--distro sles "
-          . "--early-exit "
+          # . "--early-exit "
           . "-R '$instance_id' "
           . "--region 'eu-central-1' "
           . "-u ec2-user "
           . "--ssh-private-key QA_SSH_KEY.pem "
           . "--ssh-key-name QA_SSH_KEY "
-          . "test_sles_hostname");
+          . "--results-dir ipa_results "
+          . "test_sles "
+          . "test_sles_ec2 "
+          . "test_sles_on_demand ",
+        timeout => 600
+    );
+
+    save_logs;
 }
 
 sub post_fail_hook {
+
+    save_logs;
+
+    # Terminate all instances
```

Comment:
I recommend you mention here why you want to terminate the instances. Is it to prevent them running *forever*? What about the "passed" case then?

Reply:
This is something I currently don't know. 
For now I would say a instance should only run within a testmodule and once the module is finished the instance should quit  also.
And as long as we do not have a better resource management here,  I would say using one  AWS account only once and not in parallel test-runs.

Reply:
Regarding the "passed" case. By default IPA terminates the instance if all tests run successful.